File: frontend/lib/api/batches.ts
export interface BatchItemCreate {
  garment_id: string;
  model_id: string;
  cloth_type: "upper_body" | "lower_body" | "dress";
}

export interface BatchCreateRequest {
  name?: string;
  items: BatchItemCreate[];
}

export interface BatchCreateResponse {
  id: string;
  name: string;
  status: string;
  total_items: number;
  completed_count: number;
  failed_count: number;
  created_at: string;
}

export interface BatchItemResponse {
  id: string;
  garment_id: string;
  model_id: string;
  cloth_type: string;
  status: string;
  vton_job_id: string | null;
  error_message: string | null;
  retry_count: number;
  created_at: string;
}

export interface BatchDetailResponse {
  id: string;
  name: string;
  status: string;
  total_items: number;
  completed_count: number;
  failed_count: number;
  created_at: string;
  completed_at: string | null;
  items: BatchItemResponse[];
}

export interface BatchListItem {
  id: string;
  name: string;
  status: string;
  total_items: number;
  completed_count: number;
  failed_count: number;
  created_at: string;
}

export interface BatchListResponse {
  items: BatchListItem[];
  total: number;
  page: number;
  page_size: number;
}

import { apiFetch } from "@/lib/api";

export async function createBatch(
  request: BatchCreateRequest
): Promise<BatchCreateResponse> {
  return apiFetch("/api/batches", {
    method: "POST",
    body: JSON.stringify(request),
  });
}

export async function getBatch(batchId: string): Promise<BatchDetailResponse> {
  return apiFetch(`/api/batches/${batchId}`);
}

export async function listBatches(
  page = 1,
  pageSize = 20
): Promise<BatchListResponse> {
  return apiFetch(`/api/batches?page=${page}&page_size=${pageSize}`);
}

export async function retryBatchItem(
  batchId: string,
  itemId: string
): Promise<BatchItemResponse> {
  return apiFetch(`/api/batches/${batchId}/items/${itemId}/retry`, {
    method: "POST",
  });
}

File: backend/api/schemas/batches.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field, field_validator


class BatchItemStatus(str):
    pending = "pending"
    processing = "processing"
    complete = "complete"
    failed = "failed"


class BatchJobStatus(str):
    pending = "pending"
    in_progress = "in-progress"
    complete = "complete"
    partial = "partial"
    failed = "failed"


class BatchItemCreateRequest(BaseModel):
    garment_id: uuid.UUID
    model_id: uuid.UUID
    cloth_type: str

    @field_validator("cloth_type")
    @classmethod
    def validate_cloth_type(cls, v: str) -> str:
        allowed = {"upper_body", "lower_body", "dress"}
        if v not in allowed:
            raise ValueError(
                f"cloth_type must be one of: {', '.join(sorted(allowed))}"
            )
        return v


class BatchCreateRequest(BaseModel):
    name: str | None = None
    items: list[BatchItemCreateRequest]

    @field_validator("items")
    @classmethod
    def validate_items(cls, v: list[BatchItemCreateRequest]) -> list[BatchItemCreateRequest]:
        if len(v) == 0:
            raise ValueError("Batch must contain at least 1 item")
        if len(v) > 100:
            raise ValueError("Batch size exceeds maximum of 100 items")
        return v


class BatchItemResponse(BaseModel):
    id: uuid.UUID
    garment_id: uuid.UUID
    model_id: uuid.UUID
    cloth_type: str
    status: str
    vton_job_id: uuid.UUID | None = None
    error_message: str | None = None
    retry_count: int = 0
    created_at: datetime


class BatchCreateResponse(BaseModel):
    id: uuid.UUID
    name: str
    status: str
    total_items: int
    completed_count: int = 0
    failed_count: int = 0
    created_at: datetime


class BatchDetailResponse(BaseModel):
    id: uuid.UUID
    name: str
    status: str
    total_items: int
    completed_count: int
    failed_count: int
    created_at: datetime
    completed_at: datetime | None = None
    items: list[BatchItemResponse]


class BatchListItemResponse(BaseModel):
    id: uuid.UUID
    name: str
    status: str
    total_items: int
    completed_count: int
    failed_count: int
    created_at: datetime


class BatchListResponse(BaseModel):
    items: list[BatchListItemResponse]
    total: int
    page: int
    page_size: int
